File: litho-it/src/test/java/com/facebook/litho/InterStagePropsTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static com.facebook.litho.LifecycleStep.ON_BIND;
import static com.facebook.litho.LifecycleStep.ON_MOUNT;
import static com.facebook.litho.LifecycleStep.ON_UNBIND;
import static com.facebook.litho.LifecycleStep.ON_UNMOUNT;
import static org.assertj.core.api.Java6Assertions.assertThat;

import com.facebook.litho.testing.LegacyLithoViewRule;
import com.facebook.litho.testing.testrunner.LithoTestRunner;
import com.facebook.litho.widget.MountSpecInterStagePropsTester;
import com.facebook.litho.widget.SimpleStateUpdateEmulator;
import com.facebook.litho.widget.SimpleStateUpdateEmulatorSpec;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.annotation.LooperMode;

@LooperMode(LooperMode.Mode.LEGACY)
@RunWith(LithoTestRunner.class)
public class InterStagePropsTest {

  private ComponentContext mContext;
  public final @Rule LegacyLithoViewRule mLegacyLithoViewRule = new LegacyLithoViewRule();

  @Before
  public void setUp() {
    mContext = mLegacyLithoViewRule.getContext();
    mLegacyLithoViewRule.useLithoView(new LithoView(mContext));
  }

  @Test
  public void interStageProp_FromPrepare_usedIn_OnBind() {
    final LifecycleTracker lifecycleTracker = new LifecycleTracker();
    final SimpleStateUpdateEmulatorSpec.Caller stateUpdater =
        new SimpleStateUpdateEmulatorSpec.Caller();

    final Component root = createComponent(lifecycleTracker, stateUpdater);
    mLegacyLithoViewRule.setRoot(root).attachToWindow().measure().layout();

    lifecycleTracker.reset();

    stateUpdater.increment();

    assertThat(lifecycleTracker.getSteps())
        .describedAs("On Bind should be called")
        .contains(ON_BIND);
  }

  @Test
  public void interStageProp_FromBind_usedIn_OnUnbind() {
    final LifecycleTracker lifecycleTracker = new LifecycleTracker();
    final SimpleStateUpdateEmulatorSpec.Caller stateUpdater =
        new SimpleStateUpdateEmulatorSpec.Caller();

    final Component root = createComponent(lifecycleTracker, stateUpdater);
    mLegacyLithoViewRule.setRoot(root).attachToWindow().measure().layout();

    lifecycleTracker.reset();

    mLegacyLithoViewRule.detachFromWindow();

    assertThat(lifecycleTracker.getSteps())
        .describedAs("On Unbind should be called")
        .contains(ON_UNBIND);
  }

  @Test
  public void interStageProp_FromMeasure_usedIn_OnMount() {
    final LifecycleTracker lifecycleTracker = new LifecycleTracker();
    final SimpleStateUpdateEmulatorSpec.Caller stateUpdater =
        new SimpleStateUpdateEmulatorSpec.Caller();

    final Component root = createComponent(lifecycleTracker, stateUpdater);
    mLegacyLithoViewRule.setRoot(root).attachToWindow().measure().layout();

    assertThat(lifecycleTracker.getSteps())
        .describedAs("On Mount should be called")
        .contains(ON_MOUNT);
  }

  @Test
  public void interStageProp_FromBoundsDefined_usedIn_OnUnMount() {
    final LifecycleTracker lifecycleTracker = new LifecycleTracker();
    final SimpleStateUpdateEmulatorSpec.Caller stateUpdater =
        new SimpleStateUpdateEmulatorSpec.Caller();

    final Component root = createComponent(lifecycleTracker, stateUpdater);
    mLegacyLithoViewRule.setRoot(root).attachToWindow().measure().layout();

    lifecycleTracker.reset();

    mLegacyLithoViewRule.getLithoView().unmountAllItems();

    assertThat(lifecycleTracker.getSteps())
        .describedAs("On Unmount should be called")
        .contains(ON_UNMOUNT);
  }

  private Component createComponent(
      LifecycleTracker lifecycleTracker, SimpleStateUpdateEmulatorSpec.Caller stateUpdater) {
    return Column.create(mLegacyLithoViewRule.getContext())
        .child(
            MountSpecInterStagePropsTester.create(mLegacyLithoViewRule.getContext())
                .lifecycleTracker(lifecycleTracker))
        .child(
            SimpleStateUpdateEmulator.create(mLegacyLithoViewRule.getContext())
                .caller(stateUpdater))
        .build();
  }
}

File: litho-it/src/test/java/com/facebook/litho/InterStagePropsTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import com.facebook.litho.testing.LithoTestRule
import com.facebook.litho.testing.testrunner.LithoTestRunner
import com.facebook.litho.widget.MountSpecInterStagePropsTester
import com.facebook.litho.widget.SimpleStateUpdateEmulator
import com.facebook.litho.widget.SimpleStateUpdateEmulatorSpec
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.LooperMode

@LooperMode(LooperMode.Mode.LEGACY)
@RunWith(LithoTestRunner::class)
class InterStagePropsTest {

  private lateinit var context: ComponentContext

  @JvmField @Rule val lithoViewRule = LithoTestRule()

  @Before
  fun setUp() {
    context = lithoViewRule.context
  }

  @Test
  fun interStageProp_FromPrepare_usedIn_OnBind() {
    val lifecycleTracker = LifecycleTracker()
    val stateUpdater = SimpleStateUpdateEmulatorSpec.Caller()
    val root = createComponent(lifecycleTracker, stateUpdater)
    val testLithoView = lithoViewRule.render { root }.attachToWindow().measure().layout()
    lifecycleTracker.reset()
    stateUpdater.increment()
    testLithoView.detachFromWindow()
    // We need to make sure layout happens because which triggers mount and bind together
    testLithoView.attachToWindow()
    val config = testLithoView.lithoView.configuration
    if (config != null && config.enableFixForIM) {
      testLithoView.lithoView.notifyVisibleBoundsChanged()
      assertThat(lifecycleTracker.steps)
          .describedAs("On Bind should be called")
          .containsExactly(
              LifecycleStep.ON_UNBIND,
              LifecycleStep.ON_UNMOUNT,
              LifecycleStep.ON_MOUNT,
              LifecycleStep.ON_BIND)
    } else {
      assertThat(lifecycleTracker.steps)
          .describedAs("On Bind should be called")
          .containsExactly(LifecycleStep.ON_UNBIND, LifecycleStep.ON_BIND)
    }
  }

  @Test
  fun interStageProp_FromBind_usedIn_OnUnbind() {
    val lifecycleTracker = LifecycleTracker()
    val stateUpdater = SimpleStateUpdateEmulatorSpec.Caller()
    val root = createComponent(lifecycleTracker, stateUpdater)
    val testLithoView = lithoViewRule.render { root }.attachToWindow().measure().layout()
    lifecycleTracker.reset()
    testLithoView.detachFromWindow()
    assertThat(lifecycleTracker.steps)
        .describedAs("On Unbind should be called")
        .contains(LifecycleStep.ON_UNBIND)
  }

  @Test
  fun interStageProp_FromMeasure_usedIn_OnMount() {
    val lifecycleTracker = LifecycleTracker()
    val stateUpdater = SimpleStateUpdateEmulatorSpec.Caller()
    val root = createComponent(lifecycleTracker, stateUpdater)
    lithoViewRule.render { root }.attachToWindow().measure().layout()
    assertThat(lifecycleTracker.steps)
        .describedAs("On Mount should be called")
        .contains(LifecycleStep.ON_MOUNT)
  }

  @Test
  fun interStageProp_FromBoundsDefined_usedIn_OnUnMount() {
    val lifecycleTracker = LifecycleTracker()
    val stateUpdater = SimpleStateUpdateEmulatorSpec.Caller()
    val root = createComponent(lifecycleTracker, stateUpdater)
    val testLithoView = lithoViewRule.render { root }.attachToWindow().measure().layout()
    lifecycleTracker.reset()
    testLithoView.lithoView.unmountAllItems()
    assertThat(lifecycleTracker.steps)
        .describedAs("On Unmount should be called")
        .contains(LifecycleStep.ON_UNMOUNT)
  }

  private fun createComponent(
      lifecycleTracker: LifecycleTracker,
      stateUpdater: SimpleStateUpdateEmulatorSpec.Caller
  ): Component =
      Column.create(context)
          .child(MountSpecInterStagePropsTester.create(context).lifecycleTracker(lifecycleTracker))
          .child(SimpleStateUpdateEmulator.create(context).caller(stateUpdater))
          .build()
}
